source/patterns/@aws-solutions-constructs/resources/kms-key-policy-updater/lib/index.ts

```diff
@@ -0,0 +1,109 @@
+/**
+ *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
+ *  with the License. A copy of the License is located at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
+ *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
+ *  and limitations under the License.
+ */
+
+import { KMSClient, GetKeyPolicyCommand, DescribeKeyCommand, PutKeyPolicyCommand, KeyManagerType } from "@aws-sdk/client-kms";
+
+const kmsClient = new KMSClient();
+
+export const handler = async (e: any, context: any) => {
+
+  let status = 'SUCCESS';
+  let responseData = {};
+
+  if (e.RequestType === 'Create' || e.RequestType === 'Update') {
+
+    try {
+      const kmsKeyId = e.ResourceProperties.KmsKeyId;
+      const cloudFrontDistributionId = e.ResourceProperties.CloudFrontDistributionId;
+      const accountId = e.ResourceProperties.AccountId;
+      const region = process.env.AWS_REGION;
+
+      const describeKeyCommandResponse = await kmsClient.send(new DescribeKeyCommand({
+        KeyId: kmsKeyId
+      }));
+
+      if (describeKeyCommandResponse.KeyMetadata?.KeyManager === KeyManagerType.AWS) {
+        return {
+          Status: 'SUCCESS',
+          Reason: 'An AWS managed key was provided, no action needed from the custom resource, exiting now.',
+          PhysicalResourceId: e.PhysicalResourceId ?? context.logStreamName,
+          StackId: e.StackId,
+          RequestId: e.RequestId,
+          LogicalResourceId: e.LogicalResourceId,
+          Data: 'An AWS managed key was provided, no action needed from the custom resource, exiting now.',
+        };
+      }
+
+      const getKeyPolicyCommandResponse = await kmsClient.send(new GetKeyPolicyCommand({
+        KeyId: kmsKeyId,
+        PolicyName: 'default'
+      }));
+
+      if (!getKeyPolicyCommandResponse.Policy) {
+        return {
+          Status: 'FAILED',
+          Reason: 'An error occurred while retrieving the key policy',
+          PhysicalResourceId: e.PhysicalResourceId ?? context.logStreamName,
+          StackId: e.StackId,
+          RequestId: e.RequestId,
+          LogicalResourceId: e.LogicalResourceId,
+          Data: 'An error occurred while retrieving the key policy',
+        };
+      }
+
+      const keyPolicy = JSON.parse(getKeyPolicyCommandResponse?.Policy);
+
+      // Update the existing key policy to allow the cloudfront distribution to use the key
+      keyPolicy.Statement.push({
```

Comment:
Just a note on what we discussed, this is where we should verify the policy statement hasn't already been added, otherwise risk the possibility of adding a new, duplicate statement on every stack update that results in the custom resource update invocation (which isn't necessarily every time the stack is updated).

Reply:
Roger that, added this in commit d50315298ecb4a072a128656ad57a38761b9aab6 in the form of a short-circuit that'll return success w/ no action if a duplicate key policy is detected. If not, the routine will proceed and append the new key policy to the statements array. Have a look and let me know what you think / if there are any additional concerns. Test cases also updated in the local /tests folder.